File: Exercises_Java/Alarm_System/src/impl/FireSensor.java
package impl;

import java.util.Random;

public class FireSensor implements HazardSensor {

  private String location;
  private String sensorType;
  private double batteryPercentage;
  private Random randomint;

  public FireSensor(String location) {
    this.location = location;
    this.sensorType = "Fire_sensor";
    this.batteryPercentage = 100.0;
    randomint = new Random();
  }

  /**
   * Raises an alarm 5% of the time it is called, and drains 10% battery between each poll
   *
   * @return true 5% of the time
   * Assume that the sensor can't be triggered if the battery percentage is zero
   */
  @Override
  public boolean isTriggered() {
    if (this.getBatteryPercentage() != 0) batteryPercentage -= 10;
      return randomint.nextInt(100) <= 5;
  }

  @Override
  public String getLocation() {
    return location;
  }

  @Override
  public String getSensorType() {
    return sensorType;
  }

  @Override
  public double getBatteryPercentage() {
    return batteryPercentage;
  }

}

File: exercises/week_02/src/main/scala/sensors/FireSensor.scala
package sensors

/**
  * created by LucieCBurgess on 24.01.2017
  * Task is to implement and test the FireSensor methods
  * FireSensor implements Sensor and HazardSensor
  * 5% of the time it is called, it raises an alarm
  * Drains 10% of the battery between each poll
  */

import traits.{HazardSensor, Sensor}

import scala.util.Random

class FireSensor(private val location: String, private val sensorType: String = "Fire_sensor",
                 private var batteryPercentage: Int = 100) extends Sensor with HazardSensor {

    override def isTriggered: Boolean = {
      if (getBatteryPercentage != 0) batteryPercentage -= 10
      Random.nextInt(100) < 5
    }

    override def getLocation: String = location

    override def getSensorType: String = sensorType

    override def getBatteryPercentage: Double = batteryPercentage

}
